src/main/java/picard/sam/CheckDuplicateMarking.java

```diff
@@ -0,0 +1,165 @@
+            "with the same queryname are equal.";
+
+    @Argument(doc = "Input BAM or SAM file to check.", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(doc = "Output file into which bad querynames will be placed (if not null).", shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, optional = true)
+    public File OUTPUT = null;
+
+    @Argument(doc = "Which reads of the same name should be checked to have same duplicate marking.")
+    public Mode MODE = Mode.ALL;
+
+    private String currentReadName = "";
+    private boolean currentReadDuplicateMarked = false;
+
+    private final Log log = Log.getInstance(CheckDuplicateMarking.class);
+    private final ProgressLogger progress = new ProgressLogger(log, (int) 1e6, "Checked.");
+
+    static private int NUM_WARNINGS = 100;
+    private int numBadRecords = 0;
+
+    public enum Mode implements CommandLineParser.ClpEnum {
+        ALL("Check all reads."),
+        PRIMARY_ONLY("Check primary alignments."),
+        PRIMARY_MAPPED_ONLY("Check mapped alignments."),
+        PRIMARY_PROPER_PAIR_ONLY("Check mapped alignments.");
+
+        private final String message;
+
+        Mode(final String message) {
+            this.message = message;
+        }
+
+        public String getHelpDoc() {
+            return message;
+        }
+    }
+
+    @Override
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        if (OUTPUT != null) {
+            IOUtil.assertFileIsWritable(OUTPUT);
+        }
+
+        try (SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT)) {
+
+            //sort by queryname if not already
+            final Iterable<SAMRecord> samRecordIterable;
+            if (reader.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.queryname) {
+                log.info("Input file isn't queryname sorted. Sorting into temp space.");
+
+                final Path[] tmpPaths = TMP_DIR.stream()
+                        .map(File::toPath)
+                        .toArray(Path[]::new);
+
+                final ProgressLogger sortProgress = new ProgressLogger(log, (int) 1e6, "Read into sorter");
+                final SortingCollection<SAMRecord> alignmentSorter = SortingCollection.newInstance(SAMRecord.class,
+                        new BAMRecordCodec(reader.getFileHeader()),
+                        new SAMRecordQueryNameComparator(),
+                        MAX_RECORDS_IN_RAM,
+                        tmpPaths);
+
+                for (final SAMRecord rec : reader) {
+                    alignmentSorter.add(rec);
+                    sortProgress.record(rec);
+                }
+                samRecordIterable = alignmentSorter;
+            } else {
+                samRecordIterable = reader;
+            }
+
+            checkDuplicateMarkingsInIterable(samRecordIterable);
+            log.info("Found " + numBadRecords + " bad records.");
```

Comment:
Use `error` not `info` if there are bad records, and/or only log if there's an error.